packages/@react-aria/utils/src/useSyntheticBlurEvent.ts

```diff
@@ -0,0 +1,205 @@
+    this.currentTarget = nativeEvent.currentTarget as Element;
+    this.relatedTarget = nativeEvent.relatedTarget as Element;
+    this.bubbles = nativeEvent.bubbles;
+    this.cancelable = nativeEvent.cancelable;
+    this.defaultPrevented = nativeEvent.defaultPrevented;
+    this.eventPhase = nativeEvent.eventPhase;
+    this.isTrusted = nativeEvent.isTrusted;
+    this.timeStamp = nativeEvent.timeStamp;
+    this.type = type;
+  }
+
+  isDefaultPrevented(): boolean {
+    return this.nativeEvent.defaultPrevented;
+  }
+
+  preventDefault(): void {
+    this.defaultPrevented = true;
+    this.nativeEvent.preventDefault();
+  }
+
+  stopPropagation(): void {
+    this.nativeEvent.stopPropagation();
+    this.isPropagationStopped = () => true;
+  }
+
+  isPropagationStopped(): boolean {
+    return false;
+  }
+
+  persist() {}
+}
+
+export function useSyntheticBlurEvent(onBlur: (e: ReactFocusEvent) => void) {
+  let stateRef = useRef({
+    isFocused: false,
+    onBlur,
+    observer: null as MutationObserver,
+    target: null as HTMLElement
+  });
+  stateRef.current.onBlur = onBlur;
+
+  // Clean up MutationObserver on unmount. See below.
+  // eslint-disable-next-line arrow-body-style
+  useLayoutEffect(() => {
+    const state = stateRef.current;
+    return () => {
+      if (stateRef.current.target) {
+        stateRef.current.target.removeEventListener('focusout', onBlurHandler);
+      }
+      if (state.observer) {
+        state.observer.disconnect();
+        state.observer = null;
+      }
+    };
+  }, []);
+
+
+  let onBlurHandler = useCallback((e: FocusEvent) => {
+    let target = e.target;
+    stateRef.current.isFocused = false;
+
+    if (((target instanceof HTMLButtonElement ||
+      target instanceof HTMLInputElement ||
+      target instanceof HTMLTextAreaElement ||
+      target instanceof HTMLSelectElement) && target.disabled) || !document.body.contains(e.target as HTMLElement)) {
+      // For backward compatibility, dispatch a (fake) React synthetic event.
+      stateRef.current.onBlur?.(new SyntheticBlurEvent('blur', e));
+    }
+
+    // We no longer need the MutationObserver once the target is blurred.
+    if (stateRef.current.observer) {
+      stateRef.current.observer.disconnect();
+      stateRef.current.observer = null;
+    }
+  }, [stateRef]);
+
+  // This function is called during a React onFocus event.
+  return useCallback((e: ReactFocusEvent) => {
+
+    let target = e.target as HTMLElement;
+    stateRef.current.isFocused = true;
+
+    if (stateRef.current.target && target !== stateRef.current.target) {
+      stateRef.current.target.removeEventListener('focusout', onBlurHandler);
+      domNodeRemovedObserver.removeTarget(stateRef.current.target);
+      target.addEventListener('focusout', onBlurHandler, {once: true});
+    } else if (!stateRef.current.target || e.target !== stateRef.current.target) {
+      target.addEventListener('focusout', onBlurHandler, {once: true});
+    }
```

Comment:
What scenarios is this covering? The first `else if` is for when we first call `onSyntheticFocus` and the target hasn't been set up in the stateRef via line 122 but the others are for if `stateRef.current.target` already exists? Where else is `stateRef.current.target` set?

Reply:
Ah I see, its a single instance from FocusScope so we clean up the previous focused target's blur handlers. I guess the disconnect done in line 82 is a safety net of sorts?

Reply:
i think you're conflating unmounting of the component calling this hook with the target element unmounting

the target unmounting won't cause any cleanup

so this bit of code here is to remove blur listeners that we added to a previous target whenever we *change* targets, aka we went from focusing one element to another but still within the scope. this will happen before synthetic blurring might happen, so this is the safety net more so than the other way around.

Reply:
Ok that makes more sense, so `stateRef.current.target` is the previous focused element and `target` would be the newly focused element.